File: Evolutionary Algorithms/MATLAB Codes/GA.m
clc;
clear all;

% Genetic Algorithm Parameters
pop_size = 50;
num_vars = 30;
lb = -100;
ub = 100;
num_generations = 1000;
mutation_rate = 0.1;
crossover_rate = 0.7;

% Initialize population
population = lb + (ub - lb) * rand(pop_size, num_vars);

% Main loop
for gen = 1:num_generations
    
    % Evaluate fitness for each individual in the population
    fitness = zeros(pop_size, 1);
    for i = 1:pop_size
        fitness(i) = sphere_func(population(i, :));
    end
    best_fitness = min(fitness);
    
    % Selection
    tournament_size = 2;
    tournament_winners = zeros(pop_size, num_vars);
    for i = 1:pop_size
        tournament_inds = randperm(pop_size, tournament_size);
        tournament_fitness = fitness(tournament_inds);
        [~, best_ind] = min(tournament_fitness);
        tournament_winners(i, :) = population(tournament_inds(best_ind), :);
    end
    
    % Crossover
    crossover_inds = rand(pop_size, num_vars) < crossover_rate;
    offspring = zeros(pop_size, num_vars);
    for i = 1:pop_size
        parent1 = tournament_winners(i, :);
        if i == pop_size
            parent2 = tournament_winners(1, :);
        else
            parent2 = tournament_winners(i+1, :);
        end
        offspring(i, crossover_inds(i,:)) = parent1(crossover_inds(i,:));
        offspring(i, ~crossover_inds(i,:)) = parent2(~crossover_inds(i,:));
    end
    
    % Mutation
    mutation_inds = rand(pop_size, num_vars) < mutation_rate;
    mutation_amounts = lb + (ub - lb) * rand(pop_size, num_vars);
    offspring(mutation_inds) = mutation_amounts(mutation_inds);
    
    % Update population
    population = offspring;
    
    % Display current best fitness
    fprintf('Generation %d: Best Fitness = %f\n', gen, best_fitness);
end

% Final evaluation
final_fitness = zeros(pop_size, 1);
for i = 1:pop_size
    final_fitness(i) = sphere_func(population(i, :));
end
best_solution = population(final_fitness == min(final_fitness), :);
best_fitness = min(final_fitness);
fprintf('Final Best Fitness = %f\n', best_fitness);
disp('Best Solution:');
disp(best_solution);

File: Evolutionary Algorithms/Python Codes/GA.py
# -*- coding: utf-8 -*-
"""
Created on Thu Sep  7 03:52:58 2023

@author: msi-1
"""

import random

# Genetic Algorithm parameters
POPULATION_SIZE = 100
GENERATION_COUNT = 50
CROSSOVER_RATE = 0.8
MUTATION_RATE = 0.1
CHROMOSOME_LENGTH = 4
LOWER_BOUND = -5.12
UPPER_BOUND = 5.12

# Generate a random chromosome
def generate_chromosome():
    return [random.randint(0, 1) for _ in range(CHROMOSOME_LENGTH)]

# Evaluate the fitness of an individual
def evaluate_fitness(chromosome):
    x = decode_chromosome(chromosome)
    #Here you can change the objective function value 
    fitness_value = sum([gene**2 for gene in x])
    return fitness_value

# Decode binary chromosome to real-valued representation
def decode_chromosome(chromosome):
    x = []
    for gene in chromosome:
        value = LOWER_BOUND + (UPPER_BOUND - LOWER_BOUND) * int("".join(map(str, chromosome)), 2) / (2 ** CHROMOSOME_LENGTH - 1)
        x.append(value)
    return x

# Perform selection using tournament selection
def selection(population):
    tournament_size = 5
    selected_parents = []
    for _ in range(len(population)):
        tournament = random.sample(population, tournament_size)
        winner = min(tournament, key=lambda x: x[1])
        selected_parents.append(winner[0])
    return selected_parents

# Perform crossover between two parents
def crossover(parent1, parent2):
    if random.random() < CROSSOVER_RATE:
        crossover_point = random.randint(1, CHROMOSOME_LENGTH - 1)
        child1 = parent1[:crossover_point] + parent2[crossover_point:]
        child2 = parent2[:crossover_point] + parent1[crossover_point:]
        return child1, child2
    else:
        return parent1, parent2

# Perform mutation on an individual
def mutate(individual):
    mutated_individual = individual.copy()
    for i in range(CHROMOSOME_LENGTH):
        if random.random() < MUTATION_RATE:
            mutated_individual[i] = 1 - mutated_individual[i]
    return mutated_individual

# Generate an initial population
population = [(generate_chromosome(), 0) for _ in range(POPULATION_SIZE)]

# Genetic Algorithm main loop
for _ in range(GENERATION_COUNT):
    # Evaluate fitness of each individual in the population
    population = [(chromosome, evaluate_fitness(chromosome)) for chromosome, _ in population]

    # Select parents for reproduction
    parents = selection(population)

    # Create offspring through crossover and mutation
    offspring = []
    for i in range(0, POPULATION_SIZE, 2):
        parent1 = parents[i]
        parent2 = parents[i + 1]
        child1, child2 = crossover(parent1, parent2)
        child1 = mutate(child1)
        child2 = mutate(child2)
        offspring.extend([child1, child2])

    # Replace the old population with the offspring
    population = [(chromosome, 0) for chromosome in offspring]

# Get the best individual as the solution
best_individual = min(population, key=lambda x: x[1])[0]
decoded_solution = decode_chromosome(best_individual)
fitness_value = evaluate_fitness(best_individual)
# Print the solution
print("Best Solution:", decoded_solution)
print("Best Fitness:", fitness_value)

